paddle/framework/variable.md

```diff
@@ -0,0 +1,52 @@
+# Design Doc: Variable
+
+
+Variable is also known as *blob* in MxNet and Caffe2.  It is the input and output type of operators, where a neural network is a graph of operators.
+
+## Requirements: Lazy Memory Allocation
+
+For the flexibility of a DL system, a variable should be able to contain any typed value -- a tensor in most cases, but could also be some integer IDs or a scope of other variables in the case of RNN.
+
+To use the minimum amount of memory, we'd like that a variable to allocate memory when it has to, or, lazy memory allocation.  Let's take the following example:
+
+```cpp
+Variable vr, v1, v2;
+
+Tensor* t1 = new Tensor();
+Tensor* t2 = new Tensor();
+
+Randomize(
+  /* malloc */ v1.GetMutable<Tensor>().mutable_data<float16>(DDim(100,200)),
+  /* size */ t1.Size());
+  
+Randomize(
+  /* malloc */ v2.GetMutable<Tensor>().mutable_data<float16>(DDim(200,300)),
+  /* size */ t2.Size());
+  
+Mult(
+  /*result*/ vr.GetMutable<Tensor>().mutable_data<v1.Type()>(SizeOfMult(v1, v2)),
+  /*input1*/ v1.Get<Tensor>().data(),
+  /*input2*/ v2.Get<Tensor>().data());
+```
+     
+We see that a variable holds nothing until `Variable::GetMutable<Tensor>()` allocates a tensor and puts it in the variable.  Similarly, a tensor gets its memory until `Tensor::mutable_data()`.
```

Comment:
It seems that 'lazy allocation' is split into two parts:  the part in `Variable` and the part in `Tensor`. I am wondering whether following design will be a little better:

In `Tensor`:
```cpp
Tensor<Device, T, D> {
  // return an empty tensor with ptr_ pointed to nullptr
  Tensr();
  // return a tensor with new allocated memory
  Tensor(Dim<D> size);
  // check whether ptr_ points to nullptr
  bool is_empty();
}
```

In `Variable`:
before `GetMutable<Tensor>()` is called, `Variable` is assigned with an empty `Tenosr` which is constructed by `Tensor::Tensor()`. When `GetMutable<Tensor>()` is called, it checks whether the tensor is empty, if so, construct a new tensor by `Tensor::Tensor(Dim<D> size)` and then replace the empty one with it.

The replacement costs little time because the empty tensor holds no memory.

The benefit of this design is that all implement of 'lazy allocation' is now limited in `Variable::GetMutable<Tensor>()`, and we have no need to think about it during tensor implement. It is a little like Majel's approach.